### PR TITLE
[1141] Allow multiple fields in depends_on clauses

### PR DESCRIPTION
**Pull request by jcoglan**
## Changes in this PR:

This PR adds the ability to have multiple fields in a `depends_on` clause in FDL. This is, this is now allowed:

        InvoiceFields {
          A from 'field a'
          B from 'field b'
          C from 'field c' depends_on 'field a', 'field b' {
            'one', 'apple' -> Set1
            'two', 'orange' -> Set2
          }
        }

The semantics are that for each row, the values of `field a` and `field b` will be extracted and used to look up a single row in the match table. Here, this says that if `field a` has value `one` _and_ `field b` has value `apple`, then use `Set1` as the valid values for `field c`. If `field a` is `two` _and_ `field b` is `orange`, then use `Set2`.

It is an error to use a different number of matching values than there are dependent fields. For example, these matches are both errors:

        InvoiceFields {
          A from 'field a'
          B from 'field b'
          C from 'field c' depends_on 'field a', 'field b' {
            'one', 'too', 'many' -> Set1
            'too few' -> Set2
          }
        }

All the values that were previously legal as match values in `depends_on` are still legal, but now you can use lists of them, separated by commas. There is no limitation on the number of dependent fields.

### CODE REVIEW

**Comment by jcoglan** [comment]
One thing I should add here is that I also considered a design using nested hashes. As the syntax above could result in every combination of values and therefore much repetition, for example:

```
    InvoiceFields {
      A from 'field a'
      B from 'field b'
      C from 'field c' depends_on 'field a', 'field b' {
        'one', 'apple' -> Set1Apple
        'one', 'orange' -> Set1Orange
        'one', 'lemon' -> Set1Lemon
        'one', 'cherry' -> Set1Cherry
        'two', 'apple' -> Set2Apple
        'two', 'orange' -> Set2Orange
        'two', 'lemon' -> Set2Lemon
        'two', 'cherry' -> Set2Cherry
      }
    }
```

Another possible formulation of this logic is the following:


```
    InvoiceFields {
      A from 'field a'
      B from 'field b'
      C from 'field c' depends_on 'field a', 'field b' {
        'one' -> {
          'apple' -> Set1Apple
          'orange' -> Set1Orange
          'lemon' -> Set1Lemon
          'cherry' -> Set1Cherry
        }
        'two' -> {
          'apple' -> Set2Apple
          'orange' -> Set2Orange
          'lemon' -> Set2Lemon
          'cherry' -> Set2Cherry
        }
      }
    }
```

From an implementation point of view, I think this syntax would be harder to add to the grammar, and possibly harder to implement the semantics, just because the resulting data structures are more complicated. e.g. you need to check that the mapping has the same number of levels of nesting as their are dependent fields, not just check that all the keys have the same number of elements.

One concern I had with the grammar is that I want the `dictionary` rule to only allow multi-valued keys when used with `depends_on`, not in the other places where dictionaries are used. The equivalent in this alternate syntax would be to only allow nested dictionaries with `depends_on`, and nowhere else, and I think implementing that would be more complicated.

However, if we think there's a significant usability benefit in picking the nested syntax, I'd be prepared to put some time into it.

**Comment by jcoglan** [comment]
The other usability question I had is whether there will be cases where the value of one of the columns does not matter, e.g. if you pick `one` for `field a` then any value for `field b` produces the same lookup, but if you pick `two` in `field a` then the choice for `field b` becomes important, i.e.

    InvoiceFields {
      A from 'field a'
      B from 'field b'
      C from 'field c' depends_on 'field a', 'field b' {
        'one', (anything) -> Set1
        'two', 'apple' -> Set2Apple
        'two', 'orange' -> Set2Orange
        'two', 'lemon' -> Set2Lemon
        'two', 'cherry' -> Set2Cherry
      }
    }

Because we currently have a rule that the match keys must have the same number of elements as the list of fields following `depends_on`, this would require you to list every value for `field b`, i.e.:

    InvoiceFields {
      A from 'field a'
      B from 'field b'
      C from 'field c' depends_on 'field a', 'field b' {
        'one', 'apple' -> Set1
        'one', 'orange' -> Set1
        'one', 'lemon' -> Set1
        'one', 'cherry' -> Set1
        'two', 'apple' -> Set2Apple
        'two', 'orange' -> Set2Orange
        'two', 'lemon' -> Set2Lemon
        'two', 'cherry' -> Set2Cherry
      }
    }

Some potential ways of solving this include, allowing the user to use fewer values as a key, and whatever key values are given are matched against the first N columns:


    InvoiceFields {
      A from 'field a'
      B from 'field b'
      C from 'field c' depends_on 'field a', 'field b' {
        'one' -> Set1
        'two', 'apple' -> Set2Apple
        'two', 'orange' -> Set2Orange
        'two', 'lemon' -> Set2Lemon
        'two', 'cherry' -> Set2Cherry
      }
    }

Or, you could have an explicit "match anything" operator, as in:


    InvoiceFields {
      A from 'field a'
      B from 'field b'
      C from 'field c' depends_on 'field a', 'field b' {
        'one', * -> Set1
        'two', 'apple' -> Set2Apple
        'two', 'orange' -> Set2Orange
        'two', 'lemon' -> Set2Lemon
        'two', 'cherry' -> Set2Cherry
      }
    }

I tend to prefer the latter, as the version with just `'one' -> Set1` is ambiguous: is this a mistake, is `one` supposed to match `field a` or `field b`...? An explicit `*` operator makes it clear that the partial match is intentional, and would also allow matching on the last rather than the first field, as in `*, 'apple' -> Lookup`.

**Comment by jcoglan** [comment]
I have added one more commit that implements the `*` operator, following feedback from the client.

**Comment by jcoglan** [comment]
@tahb The ingestion and templates are the big outstanding questions for me. The tests of this functionality verify that the parser works, and that the validation rules it generates work, but don't speak about where the input data comes from or how templates are generated.

We should test this out with real data on staging, so would be good to get this merged asap.

**Comment by jcoglan** [comment]
@tahb Timur informs me the XLS templates are hand-built, not generated from FDL.

**Comment by jcoglan** [comment]
We have also resolved that nested hashes are not essential, and would require significant complexity to get right. I think the `*` operator is a better way of alleviating the pain of repetitive definitions.

File: app/models/framework/definition/ast/semantic_checker.rb [hunk 1]
```diff
@@ -57,13 +57,14 @@ def raise_when_dependent_reference_has_mismatched_arity(field)
...
-                  "#{key.size} values #{format_string_list(key)}"
+                  "#{format_list(field.dependent_fields)} but contains a match on " \
+                  "#{key.size} values #{format_list(key)}"
           end
         end
 
-        def format_string_list(array)
-          '(' + array.map { |value| "'#{value}'" }.join(', ') + ')'
+        def format_list(array)
+          values = array.map { |value| value.is_a?(String) ? "'#{value}'" : value }
```
Comment (tahb) [inline app/models/framework/definition/ast/semantic_checker.rb]: Is this change needed because `value` may be an instance of `Any` now?

If so, since `Any` responds to `to_s` will it not be called automatically by the interpolation?

**Comment by tahb** [comment]
Looks good, small question on a formatter. I certainly agree with the use of `*` to represent the notion of any argument as it makes it easier to understand and to account for in our logic.

File: app/models/framework/definition/ast/semantic_checker.rb (hunk 1 above)
Comment (jcoglan) [reply app/models/framework/definition/ast/semantic_checker.rb]: It would be, but I wasn't sure how to get the single quotes around string values but not around the `*` from the `Any`. Is there a better way of doing this?
